File: smart-contract/deploy.py
"""
ChainGrant Deploy Script
========================
Run this after Python + requirements are installed:

    cd smart-contract
    pip install -r requirements.txt
    algokit compile python contract.py --out-dir ./artifacts
    python deploy.py

It will print the App ID to paste into src/lib/algorand.ts
"""

import base64
import json
import os
import sys

from algosdk import account, mnemonic
from algosdk.v2client import algod
from algosdk.transaction import (
    ApplicationCreateTxn,
    PaymentTxn,
    StateSchema,
    wait_for_confirmation,
    assign_group_id,
)

# ------------------------------------------------------------------ #
# Config                                                               #
# ------------------------------------------------------------------ #

ALGOD_ADDRESS = "https://testnet-api.algonode.cloud"
ALGOD_TOKEN   = ""          # Algonode is public, no token needed

# Read sponsor mnemonic from .env or prompt
MNEMONIC = os.environ.get("SPONSOR_MNEMONIC") or input("Paste your Pera Wallet mnemonic (25 words): ").strip()

ARTIFACTS_DIR = os.path.join(os.path.dirname(__file__), "artifacts")

# ------------------------------------------------------------------ #
# Helpers                                                              #
# ------------------------------------------------------------------ #

def load_artifacts() -> tuple[bytes, bytes]:
    """Load compiled approval and clear programs."""
    approval_path = os.path.join(ARTIFACTS_DIR, "ChainGrant.approval.compiled")
    clear_path    = os.path.join(ARTIFACTS_DIR, "ChainGrant.clear.compiled")

    if not os.path.exists(approval_path):
        print(f"\n❌  Compiled artifact not found at {approval_path}")
        print("   Run: algokit compile python contract.py --out-dir ./artifacts")
        sys.exit(1)

    with open(approval_path, "rb") as f:
        approval = f.read()
    with open(clear_path, "rb") as f:
        clear = f.read()

    return approval, clear


def deploy_contract(client: algod.AlgodClient, private_key: str, sender: str) -> int:
    """Deploy the ChainGrant contract and return the App ID."""

    approval, clear = load_artifacts()

    # Global state: 4 byte-slices (sponsor, student) + 2 uint64 (total_locked, released)
    global_schema = StateSchema(num_uints=2, num_byte_slices=2)
    local_schema  = StateSchema(num_uints=0, num_byte_slices=0)

    params = client.suggested_params()

    txn = ApplicationCreateTxn(
        sender=sender,
        sp=params,
        on_complete=0,          # NoOp (create)
        approval_program=approval,
        clear_program=clear,
        global_schema=global_schema,
        local_schema=local_schema,
    )

    signed = txn.sign(private_key)
    tx_id  = client.send_transaction(signed)
    result = wait_for_confirmation(client, tx_id, 4)

    app_id: int = result["application-index"]
    return app_id


# ------------------------------------------------------------------ #
# Main                                                                 #
# ------------------------------------------------------------------ #

if __name__ == "__main__":
    # Recover keys
    private_key = mnemonic.to_private_key(MNEMONIC)
    sender      = account.address_from_private_key(private_key)

    # Connect to Testnet
    client = algod.AlgodClient(ALGOD_TOKEN, ALGOD_ADDRESS)

    print(f"\n🚀  Deploying ChainGrant contract from: {sender}")
    app_id = deploy_contract(client, private_key, sender)

    print(f"\n✅  Contract deployed successfully!")
    print(f"    App ID : {app_id}")
    print(f"    App Addr: {account.address_from_private_key(private_key)}")
    print(f"\n👉  Copy this App ID into src/lib/algorand.ts → CHAIN_GRANT_APP_ID")
    print(f"    https://lora.algokit.io/testnet/application/{app_id}")

File: smart-contract/deploy.mjs
/**
 * ChainGrant Deploy Script (Node.js)
 * =====================================
 * Uses the algosdk npm package (already in the project) to deploy
 * the compiled TEAL contract to Algorand Testnet.
 *
 * Run from the project root:
 *   node smart-contract/deploy.mjs
 */

import algosdk from 'algosdk';
import fs from 'fs';
import path from 'path';
import readline from 'readline';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// -------------------------------------------------- //
//  Config                                             //
// -------------------------------------------------- //
const ALGOD_SERVER = 'https://testnet-api.algonode.cloud';
const ALGOD_TOKEN = '';
const ALGOD_PORT = 443;

const ARTIFACTS = path.join(__dirname, 'artifacts');

// -------------------------------------------------- //
//  Helpers                                            //
// -------------------------------------------------- //

function prompt(question) {
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    return new Promise(resolve => rl.question(question, ans => { rl.close(); resolve(ans.trim()); }));
}

async function compileTeal(client, tealSource) {
    const result = await client.compile(tealSource).do();
    return new Uint8Array(Buffer.from(result.result, 'base64'));
}

// -------------------------------------------------- //
//  Main                                               //
// -------------------------------------------------- //

async function main() {
    const client = new algosdk.Algodv2(ALGOD_TOKEN, ALGOD_SERVER, ALGOD_PORT);

    // 1. Load TEAL source files
    const approvalPath = path.join(ARTIFACTS, 'ChainGrant.approval.teal');
    const clearPath = path.join(ARTIFACTS, 'ChainGrant.clear.teal');

    if (!fs.existsSync(approvalPath)) {
        console.error(`\n❌  Approval TEAL not found at ${approvalPath}`);
        console.error('   Run: algokit compile python contract.py --out-dir ./artifacts');
        process.exit(1);
    }

    const approvalTeal = fs.readFileSync(approvalPath, 'utf-8');
    const clearTeal = fs.readFileSync(clearPath, 'utf-8');

    // 2. Compile TEAL to bytecode
    console.log('\n⚙️   Compiling TEAL...');
    const approvalProgram = await compileTeal(client, approvalTeal);
    const clearProgram = await compileTeal(client, clearTeal);
    console.log('✅  TEAL compiled.');

    // 3. Get mnemonic from user
    const mn = await prompt('\n🔑  Paste your Pera Wallet mnemonic (25 words, space-separated):\n> ');
    const privateKey = algosdk.mnemonicToSecretKey(mn);
    const senderAddr = privateKey.addr;
    console.log(`\n📤  Deploying from: ${senderAddr}`);

    // 4. Build ApplicationCreateTxn
    const params = await client.getTransactionParams().do();

    const txn = algosdk.makeApplicationCreateTxnFromObject({
        sender: senderAddr,
        suggestedParams: params,
        onComplete: algosdk.OnApplicationComplete.NoOpOC,
        approvalProgram,
        clearProgram,
        numGlobalByteSlices: 3,   // sponsor (address), student (address), latest_proof_hash (string)
        numGlobalInts: 3,         // total_locked, released, initialized
        numLocalByteSlices: 0,
        numLocalInts: 0,
    });

    // 5. Sign and send
    const signedTxn = txn.signTxn(privateKey.sk);
    const sendResult = await client.sendRawTransaction(signedTxn).do();
    const txid = sendResult.txid;   // algosdk v3 returns lowercase 'txid'
    console.log(`\n⏳  Waiting for confirmation (tx: ${txid})...`);

    const result = await algosdk.waitForConfirmation(client, txid, 10);
    // algosdk v3 uses camelCase
    const appId = result.applicationIndex ?? result['application-index'];

    // 6. Done!
    console.log('\n🎉  ============================================================');
    console.log(`    CONTRACT DEPLOYED SUCCESSFULLY`);
    console.log(`    App ID   : ${appId}`);
    console.log(`    App Addr : ${algosdk.getApplicationAddress(appId)}`);
    console.log(`    Explorer : https://lora.algokit.io/testnet/application/${appId}`);
    console.log('    ============================================================');
    console.log('\n👉  Now open  src/lib/algorand.ts');
    console.log(`    Replace:  export const CHAIN_GRANT_APP_ID = 0;`);
    console.log(`    With:     export const CHAIN_GRANT_APP_ID = ${appId};`);
}

main().catch(err => {
    console.error('\n❌  Deploy failed:', err.message ?? err);
    process.exit(1);
});
